tests: scripts: add west debug support in twister

hakehuang opened the pull request:
in some multi-core platform, `west flash` can not support all core. `west debug` is supported to test secondary core, add this to twister such as with
`--west-flash-cmd=debug` in twister command
or with
`--west-flash-cmd: debug` in hwmap.yaml

and same with pytest

twister command line setting will override the one in hwamp

hakehuang commented:
"No space left on device" CI failure

gchwier reviewed (commented):
It might be better to move the changes related to `west` into a separate PR.

hakehuang commented:
need wait https://github.com/zephyrproject-rtos/zephyr/pull/100924 to merge

pdgendt commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -587,10 +587,15 @@ def _create_command(self, runner, hardware):
+        # notify west to use debug batch mode
+        if command[-1] == "debug":
+            command_extra_args.append('--batch')
+        command.append("--no-rebuild", "-d", self.build_dir)
```
This does not work

```shell
TypeError: list.append() takes exactly one argument (3 given)
```

```suggestion
        command.extend(["--no-rebuild", "-d", self.build_dir])
```

hakehuang replied:
good catch

gchwier reviewed (commented):
`--west-cmd` is not passed to Pytest, 
check method `_generate_parameters_for_hardware` in harness.py

gchwier commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -270,6 +270,17 @@ def add_parse_arguments(parser = None) -> argparse.ArgumentParser:
 
+    parser.add_argument(
+        "--west-cmd", choices=['flash', 'debug'],
+        default='flash',
```
you can remove the default here,
in handlers.py:594:
`west_cmd = self.options.west_cmd or hardware.west_cmd or "flash"`
in harness.py:483:
```
        if west_cmd := hardware.west_cmd or options.west_cmd:
            command.append(f'--west-cmd={west_cmd}')
```

gchwier commented on scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py:
add here:
```
        if self.device_config.west_cmd == 'debug':
            command_extra_args.append('--batch')
```

tested with command:
`./scripts/twister -vv -ll debug -T samples/subsys/testsuite/pytest/shell -s sample.pytest.shell --device-testing --hardware-map ~/tmp/hardware_map.yml --west-flash --no-clean --west-cmd debug`

hakehuang commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -587,10 +587,15 @@ def _create_command(self, runner, hardware):
+        # notify west to use debug batch mode
+        if command[-1] == "debug":
+            command_extra_args.append('--batch')
+        command.extend(["--no-rebuild", "-d", self.build_dir])
```
this is intentional, if the hwmap set the debug mode, it will override others.

gchwier commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -270,6 +270,17 @@ def add_parse_arguments(parser = None) -> argparse.ArgumentParser:
 
+    parser.add_argument(
+        "--west-cmd", choices=['flash', 'debug'],
+        default=None,
```
this line can be removed

pdgendt commented on scripts/pylib/pytest-twister-harness/src/twister_harness/plugin.py:
```diff
@@ -139,6 +139,10 @@ def pytest_addoption(parser: pytest.Parser):
         '--extra-test-args',
         help='Additional args passed to the test binary'
     )
+    twister_harness_group.addoption(
```
Add explicit `choices=('flash', 'debug')` instead of only describing them?

pdgendt reviewed (commented):
> @pdgendt I update due to some missing in the pytest harness support, and refine the code as @gchwier recommends, Please take a look. now all test with twister/west/pytest have pass. Thanks in advance.

It does need assignee approval anyway..

hakehuang commented:
@nashif please

hakehuang committed:
tests: scripts: add west debug support in twister

in some multi-core platform, `west flash` can not support all cores.
`west debug` is supported to test secondary core, add this to twister
such as with
`--west-flash-cmd=debug` in twister command
or with
`--west-flash-cmd: debug` in hwmap.yaml

and same with pytest

command line setting will overrid the one in hwmap

Signed-off-by: Hake Huang <hake.huang@nxp.com>
Co-authored-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

hakehuang committed:
tests: twister: add west_flash_cmd mock

add west_flash_cmd for debug testing

Signed-off-by: Hake Huang <hake.huang@nxp.com>

pdgendt commented on scripts/pylib/pytest-twister-harness/src/twister_harness/plugin.py:
```diff
@@ -139,6 +139,11 @@ def pytest_addoption(parser: pytest.Parser):
+    twister_harness_group.addoption(
+        '--west-flash-cmd',
+        choices=('flash', 'debug'),
+        help='west flash command flash | debug'
```
The help here isn't really helpful?
